### PR TITLE
Support ZIP 317 fees for legacy wallet operations

### LINKED ISSUE #6553: Support ZIP 317 fees in CreateTransaction
Requirements:
- always pay at least GetFeeForRelay (calculated using nTxBytes and minrelaytxfee)
- never pay more than maxtxfee
- legacy apis only: if paytxfee is set, use that subject to constraints above (and 1000-byte minimum to retain old behaviour)
- new apis: use the explicit fee if present, subject to constraints above
- otherwise pay the conventional fee subject to constraints above

Nice to have:
- get rid of the estimation logic
- remove config options that no longer make sense (txconfirmtarget, mintxfee)

### PR DESCRIPTION
**Pull request by sellout**
This also handles fee bounds (minRelayFee, maxTxFee) in WalletTxBuilder.

Fixes #6553.

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -1950,16 +1950,17 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
     bool CommitTransaction(CWalletTx& wtxNew, std::optional<std::reference_wrapper<CReserveKey>> reservekey, CValidationState& state);
 
     static CFeeRate minTxFee;
+
+    /** Prevent user from paying a non-sense fee (like 1 satoshi): 0 < fee < minRelayFee
+     * But always obey the maximum
```
Comment (nuttycom) [inline src/wallet/wallet.h:1951]:
```suggestion
    /** Adjust the requested fee by bounding it below to the minimum relay fee required
     * for a transaction of the given size and bounding it above to the maximum fee
     * configured using the `-maxtxfee` configuration option.
```

**Review by nuttycom — APPROVED** [review]
utACK with a minor doc suggestion

File: doc/release-notes.md [hunk 2]
```diff
@@ -25,11 +25,17 @@ RPC Changes
...
 - The `AllowRevealedRecipients` privacy policy is now required in order to choose a
-  transparent change address for a transaction. This will only occur when the wallet 
-  is unable to construct the transaction without selecting funds from the transparent 
+  transparent change address for a transaction. This will only occur when the wallet
+  is unable to construct the transaction without selecting funds from the transparent
   pool, so the impact of this change is that for such transactions, the user must specify
   `AllowFullyTransparent`.
 
+Removal of `-mintxfee`
+-----------
```
Comment (daira) [inline doc/release-notes.md:34]:
```suggestion
----------------------
```

Comment (sellout) [reply doc/release-notes.md:34]: I noticed that the other lines are short, too (which is why I figured we just did a set length) – should I extend all of them?

Comment (sellout) [reply doc/release-notes.md:34]: Nevermind – it’s just that I view Markdown in a proportional font, so they didn’t seem to align.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -5828,32 +5822,22 @@ bool CWallet::CommitTransaction(CWalletTx& wtxNew, std::optional<std::reference_
...
 }
 
-CAmount CWallet::GetMinimumFee(unsigned int nTxBytes, unsigned int nConfirmTarget, const CTxMemPool& pool)
+CAmount CWallet::GetMinimumFee(const CTransaction& tx, unsigned int nTxBytes)
 {
     // payTxFee is user-set "I want to pay this much"
     CAmount nFeeNeeded = payTxFee.GetFee(nTxBytes);
     // user selected total at least (default=true)
     if (fPayAtLeastCustomFee && nFeeNeeded > 0 && nFeeNeeded < payTxFee.GetFeePerK())
         nFeeNeeded = payTxFee.GetFeePerK();
```
Comment (daira) [inline src/wallet/wallet.cpp:5836]:
```suggestion
    CAmount nFeeNeeded = payTxFee.GetFee(std::min(1000, nTxBytes));
```
Proof that this is equivalent:
* `fPayAtLeastCustomFee` is initialized to `true` and nothing ever changes it.
* `payTxFee` is nonnegative and `nTxBytes` is positive. Therefore `nFeeNeeded` is nonnegative, and can only be zero if `payTxFee` is zero.
* if `payTxFee` is zero then `payTxFee.GetFee(std::min(1000, nTxBytes)) == payTxFee.GetFee(nTxBytes)`.
* otherwise, the `if` statement is equivalent to
  ```c++
  if (nFeeNeeded < payTxFee.GetFee(1000))
      nFeeNeeded = payTxFee.GetFee(1000);
  ```
  and since `GetFee` is an increasing function, this is equivalent to the suggestion.

Comment (sellout) [reply src/wallet/wallet.cpp:5836]: Shouldn’t this use `max` instead of `min`?

File: src/wallet/asyncrpcoperation_common.cpp [hunk 4]
```diff
@@ -179,6 +179,15 @@ void ThrowInputSelectionError(
...
         },
+        [](const MaxFeeError& err) {
+            throw JSONRPCError(
+                RPC_INVALID_PARAMETER,
+                strprintf(
+                    "Fee %s is greater than the maximum fee allowed by this instance (%s). Run "
+                    "zcashd with `-maxtxfee` to adjust this limit.",
+                    FormatMoney(err.fixedFee),
+                    FormatMoney(maxTxFee)));
+        },
```
Comment (daira) [inline src/wallet/asyncrpcoperation_common.cpp:190]: The default `-maxtxfee` is 0.1 ZEC (`DEFAULT_TRANSACTION_MAXFEE` in `main.h`). This seems reasonable because it corresponds to a ZIP 317 fee with 10000000/5000 = 2000 logical actions.

Comment (sellout) [reply src/wallet/asyncrpcoperation_common.cpp:190]: I agree that it’s reasonable, but with maximum prioritisation, it’s “only” 500 logical actions.

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -6746,11 +6728,7 @@ bool CWallet::ParameterInteraction(const CChainParams& params)
 {
     if (mapArgs.count("-mintxfee"))
     {
-        CAmount n = 0;
-        if (ParseMoney(mapArgs["-mintxfee"], n) && n > 0)
-            CWallet::minTxFee = CFeeRate(n);
-        else
-            return UIError(AmountErrMsg("mintxfee", mapArgs["-mintxfee"]));
+        UIWarning(_("The argument -mintxfee is no longer supported."));
     }
```
Comment (daira) [inline src/wallet/wallet.cpp:6732]: I think the check on `maxTxFee` at lines 6755-6759 no longer makes sense. I suggest requiring `maxTxFee` to be at least `conventional_fee(10)` to allow 10-action transactions at the ZIP 317 fee.

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (sellout) [reply src/wallet/wallet.cpp:5836]: Also, is the 1000 simply 1 KiB, or is there a deeper significance that would justify naming that constant?

Also, what about `CAmount nFeeNeeded = std::max(payTxFee.GetFeePerK(), payTxFee.GetFee(nTxBytes));`, which might make the 1 KiB thing clearer. Or
```c++
constexpr uint8_t B = 1;
constexpr uint8_t KiB = 1000 * B;
...
CAmount nFeeNeeded = payTxFee.GetFee(std::min(1*KiB, nTxBytes));
```

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -6746,11 +6728,7 @@ bool CWallet::ParameterInteraction(const CChainParams& params)
...
     if (mapArgs.count("-paytxfee"))
```
Comment (daira) [inline src/wallet/wallet.cpp:6733]: In order for fee rate calculations not to overflow, we must have `nSatoshisPerK <= INT64_MAX / MAX_BLOCK_SIZE`. `ParseMoney` does not ensure that limit on its result (surprisingly to me, it does not even ensure a limit of `MAX_MONEY`).

File: src/wallet/asyncrpcoperation_common.cpp (hunk 4 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_common.cpp:190]: Keep in mind that prioritization only really starts to become relevant when blocks are full of transactions paying the conventional fee, and people then decide to give up additional privacy by sending larger fees to get an advantage. Champagne problem, let's revisit that constant after 5.7.0 or something. :)

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (daira) [reply src/wallet/wallet.cpp]: > Shouldn’t this use max instead of min?

Yes, that was an error that I fixed concurrently with your comment.

> Also, is the 1000 simply 1 KiB, or is there a deeper significance that would justify naming that constant?

1000 is the fee rate denominator. It doesn't have a named constant anywhere else it is used.

Comment (sellout) [reply src/wallet/wallet.cpp]: Oh right, I reversed my units 🤦🏼

File: src/util/moneystr.cpp [hunk 7]
```diff
@@ -33,6 +33,10 @@ std::string FormatMoney(const CAmount& n)
     return str;
 }
 
+std::string DisplayMoney(const CAmount& zat)
+{
+    return CURRENCY_UNIT + " " + FormatMoney(zat);
```
Comment (nuttycom) [inline src/util/moneystr.cpp:38]:
```suggestion
    return FormatMoney(zat) + " " + CURRENCY_UNIT;
```

File: qa/rpc-tests/show_help.py [hunk 8]
```diff
@@ -90,7 +90,7 @@
        Keep at most <n> unconnectable transactions in memory (default: 100)
 
   -par=<n>
-       Set the number of script verification threads (IGNORE_NONDETERMINISTIC, 0 = auto, <0 =
+       Set the number of script verification threads (-8 to 16, 0 = auto, <0 =
```
Comment (nuttycom) [inline qa/rpc-tests/show_help.py:93]: Why is this change here?

**Review by nuttycom — COMMENTED** [review]
A couple of minor requests.

File: src/util/moneystr.cpp (hunk 7 above)
Comment (sellout) [reply src/util/moneystr.cpp:38]: Ah, it’s _this_ PR 😆 Thanks!

File: qa/rpc-tests/show_help.py (hunk 8 above)
Comment (sellout) [reply qa/rpc-tests/show_help.py]: Shouldn’t be.

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -6753,13 +6718,15 @@ bool CWallet::ParameterInteraction(const CChainParams& params)
         if (nMaxFee > HIGH_MAX_TX_FEE)
             UIWarning(_("-maxtxfee is set to a very high fee rate! Fee rates this large could be paid on a single transaction."));
         maxTxFee = nMaxFee;
-        if (CFeeRate(maxTxFee, 1000) < ::minRelayTxFee)
+        if (maxTxFee < CalculateConventionalFee(LOW_LOGICAL_ACTIONS))
```
Comment (str4d) [inline src/wallet/wallet.cpp:6721]: The removed check was definitely an error (the node config wins out over the wallet config), while the new check should really just be a warning (if the user wants to constrain their maximum fee to be below `LOW_LOGICAL_ACTIONS`, that just constrains the kinds of transactions they can create; it does not prevent them from creating transactions). So we should keep the previous error check as-is, and instead introduce this new check as a `UIWarning`.

**Review by str4d — CHANGES_REQUESTED** [review]
utACK after the error is reduced to a warning.

File: src/wallet/wallet.cpp (hunk 9 above)
Comment (str4d) [reply src/wallet/wallet.cpp:6721]: Deferred to #6577 in the interest of unblocking concurrent PRs.
